src/metabase/lib/schema/expression/temporal.cljc

```diff
@@ -4,12 +4,62 @@
    [metabase.lib.schema.mbql-clause :as mbql-clause]
    [metabase.lib.schema.temporal-bucketing :as temporal-bucketing]))
 
+(mbql-clause/define-tuple-mbql-clause :interval
+  :int
+  ::temporal-bucketing/unit.date-time.interval)
+
+;; FIXME Interval produces a fixed type but we have no :type/Interval
+(defmethod expression/type-of* :interval
+  [[_tag _opts n _unit]]
+  (expression/type-of n))
+
 ;;; TODO -- we should constrain this so that you can only use a Date unit if expr is a date, etc.
-(mbql-clause/define-tuple-mbql-clause :datetime-add
-  #_expr   [:ref ::expression/temporal]
-  #_amount :int
-  #_unit   [:ref ::temporal-bucketing/unit.date-time.interval])
-
-(defmethod expression/type-of* :datetime-add
-  [[_tag _opts expr _amount _unit]]
-  (expression/type-of expr))
+(doseq [op [:datetime-add :datetime-subtract]]
+  (mbql-clause/define-tuple-mbql-clause op
+    #_expr   [:ref ::expression/temporal]
+    #_amount :int
+    #_unit   [:ref ::temporal-bucketing/unit.date-time.interval])
+
+  (expression/register-type-of-first-arg op))
+
+(doseq [op [:get-year :get-month :get-day :get-hour :get-minute :get-second :get-quarter]]
+  (mbql-clause/define-tuple-mbql-clause op :- :type/Integer
+    [:schema [:ref ::expression/temporal]]))
+
+(mbql-clause/define-tuple-mbql-clause :datetime-diff :- :type/Integer
+  #_:datetime1 [:schema [:ref ::expression/temporal]]
+  #_:datetime2 [:schema [:ref ::expression/temporal]]
+  #_:unit [:enum "year" "month" "day" "hour" "second" "millisecond" "quarter"])
+
+(mbql-clause/define-tuple-mbql-clause :get-week :- :type/Integer
+  #_:datetime [:schema [:ref ::expression/temporal]]
+  ;; TODO should this be in the options map?
+  #_:mode [:maybe [:enum :iso :us :instance]])
+
+(mbql-clause/define-tuple-mbql-clause :convert-timezone
+  #_:datetime [:schema [:ref ::expression/temporal]]
+  ;; TODO could be better specified - perhaps with a build time macro to inline the timezones?
```

Comment:
This would be difficult to generate if we ask java.time or moment if a timezone string is valid.

Reply:
Generating Java timezone strings is not difficult (I have an implementation somewhere), I'm not sure about moment.